### ISSUE TITLE
Solving error / weirdness with mismatched constraints

### ISSUE THREAD

**Issue by gforsyth** [mamba-org/mamba#1369]
TLDR: `mamba` proposes two different environments depending on the order of the packages on the command line.

I stumbled across this while trying to install `mkdocs`, `flake8` and `python=3.10` in a new environment.

Note that `mkdocs` wants `importlib-metadata > 3.10`

https://github.com/conda-forge/mkdocs-feedstock/blob/dedbebc5265ea266c88602f464c85eec9fab1a13/recipe/meta.yaml#L26

while `flake8` wants `importlib-metadata < 4.3`

https://github.com/conda-forge/flake8-feedstock/blob/3ba564b093099800d01678c9a9cb53f4bcd49a26/recipe/meta.yaml#L28

This _seems_ like it should be resolvable but actually it isn't, because Python 3.10 requires `importlib-metadata 4.4` (technically it doesn't require it at all, but you know...)

```
mamba --version
mamba 0.19.1
conda 4.10.3
```

```
mamba create -n mkdocsflaketest mkdocs flake8 python=3.10
```
<details>

```
Looking for: ['mkdocs', 'flake8', 'python=3.10']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache
Transaction

  Prefix: /home/gil/mambaforge/envs/mkdocsflaketest

  Updating specs:

   - mkdocs
   - flake8
   - python=3.10


  Package                 Version  Build               Channel                    Size
────────────────────────────────────────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex             0.1  conda_forge         conda-forge/linux-64     Cached
  + _openmp_mutex             4.5  1_gnu               conda-forge/linux-64     Cached
  + argh                   0.26.2  pyh9f0ad1d_1002     conda-forge/noarch       Cached
  + bzip2                   1.0.8  h7f98852_4          conda-forge/linux-64     Cached
  + ca-certificates     2021.10.8  ha878542_0          conda-forge/linux-64     Cached
  + click                   8.0.3  py310hff52083_1     conda-forge/linux-64     Cached
  + flake8                  4.0.1  pyhd8ed1ab_0        conda-forge/noarch        83 KB
  + ghp-import              2.0.2  pyhd8ed1ab_0        conda-forge/noarch        15 KB
  + importlib-metadata     4.10.0  py310hff52083_0     conda-forge/linux-64      33 KB
  + jinja2                  3.0.3  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + ld_impl_linux-64       2.36.1  hea4e1c9_2          conda-forge/linux-64     Cached
  + libffi                  3.4.2  h7f98852_5          conda-forge/linux-64     Cached
  + libgcc-ng              11.2.0  h1d223b6_11         conda-forge/linux-64     Cached
  + libgomp                11.2.0  h1d223b6_11         conda-forge/linux-64     Cached
  + libnsl                  2.0.0  h7f98852_0          conda-forge/linux-64     Cached
  + libuuid                2.32.1  h7f98852_1000       conda-forge/linux-64     Cached
  + libzlib                1.2.11  h36c2ea0_1013       conda-forge/linux-64     Cached
  + markdown                3.3.6  pyhd8ed1ab_0        conda-forge/noarch        67 KB
  + markupsafe              2.0.1  py310h6acc77f_1     conda-forge/linux-64     Cached
  + mccabe                  0.6.1  py_1                conda-forge/noarch       Cached
  + mergedeep               1.3.4  pyhd8ed1ab_0        conda-forge/noarch         9 KB
  + mkdocs                  1.2.3  pyhd8ed1ab_0        conda-forge/noarch         6 MB
  + ncurses                   6.2  h58526e2_4          conda-forge/linux-64     Cached
  + openssl                 3.0.0  h7f98852_2          conda-forge/linux-64     Cached
  + packaging                21.3  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pip                    21.3.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pycodestyle             2.8.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pyflakes                2.4.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pyparsing               3.0.6  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + python                 3.10.1  h543edf9_2_cpython  conda-forge/linux-64      30 MB
  + python-dateutil         2.8.2  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + python_abi               3.10  2_cp310             conda-forge/linux-64     Cached
  + pyyaml                    6.0  py310h6acc77f_3     conda-forge/linux-64     Cached
  + pyyaml-env-tag            0.1  pyhd8ed1ab_0        conda-forge/noarch         7 KB
  + readline                  8.1  h46c0cb4_0          conda-forge/linux-64     Cached
  + setuptools             60.2.0  py310hff52083_0     conda-forge/linux-64     Cached
  + six                    1.16.0  pyh6c4a22f_0        conda-forge/noarch       Cached
  + sqlite                 3.37.0  h9cd32fc_0          conda-forge/linux-64     Cached
  + tk                     8.6.11  h27826a3_1          conda-forge/linux-64     Cached
  + tzdata                  2021e  he74cb21_0          conda-forge/noarch       Cached
  + watchdog                2.1.6  py310hff52083_1     conda-forge/linux-64      95 KB
  + wheel                  0.37.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + xz                      5.2.5  h516909a_1          conda-forge/linux-64     Cached
  + yaml                    0.2.5  h7f98852_2          conda-forge/linux-64      87 KB
  + zipp                    3.6.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + zlib                   1.2.11  h36c2ea0_1013       conda-forge/linux-64     Cached

  Summary:

  Install: 46 packages

  Total download: 36 MB

────────────────────────────────────────────────────────────────────────────────────────

Confirm changes: [Y/n] 
```

</details>

Note that `mkdocs` comes first, and `mamba` grabs `importlib-metadata 4.10.0`

conversely:

```
$ mamba create -n mkdocsflaketest flake8 mkdocs python=3.10
```

<details>

```
Looking for: ['flake8', 'mkdocs', 'python=3.10']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache
Transaction

  Prefix: /home/gil/mambaforge/envs/mkdocsflaketest

  Updating specs:

   - flake8
   - mkdocs
   - python=3.10


  Package                  Version  Build               Channel                    Size
─────────────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex              0.1  conda_forge         conda-forge/linux-64     Cached
  + _openmp_mutex              4.5  1_gnu               conda-forge/linux-64     Cached
  + bzip2                    1.0.8  h7f98852_4          conda-forge/linux-64     Cached
  + ca-certificates      2021.10.8  ha878542_0          conda-forge/linux-64     Cached
  + click                    8.0.3  py310hff52083_1     conda-forge/linux-64     Cached
  + colorama                 0.4.4  pyh9f0ad1d_0        conda-forge/noarch       Cached
  + flake8                   4.0.1  pyhd8ed1ab_1        conda-forge/noarch       Cached
  + future                  0.18.2  py310hff52083_4     conda-forge/linux-64     727 KB
  + importlib-metadata       3.3.0  pyhd8ed1ab_1        conda-forge/noarch        12 KB
  + jinja2                   3.0.3  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + joblib                   1.1.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + ld_impl_linux-64        2.36.1  hea4e1c9_2          conda-forge/linux-64     Cached
  + libffi                   3.4.2  h7f98852_5          conda-forge/linux-64     Cached
  + libgcc-ng               11.2.0  h1d223b6_11         conda-forge/linux-64     Cached
  + libgomp                 11.2.0  h1d223b6_11         conda-forge/linux-64     Cached
  + libnsl                   2.0.0  h7f98852_0          conda-forge/linux-64     Cached
  + libuuid                 2.32.1  h7f98852_1000       conda-forge/linux-64     Cached
  + libzlib                 1.2.11  h36c2ea0_1013       conda-forge/linux-64     Cached
  + livereload               2.6.3  pyh9f0ad1d_0        conda-forge/noarch        24 KB
  + lunr                     0.5.8  pyhd8ed1ab_2        conda-forge/noarch       999 KB
  + markdown                 3.3.4  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + markupsafe               2.0.1  py310h6acc77f_1     conda-forge/linux-64     Cached
  + mccabe                   0.6.1  py_1                conda-forge/noarch       Cached
  + mkdocs                   1.1.2  pyhd8ed1ab_0        conda-forge/noarch         6 MB
  + ncurses                    6.2  h58526e2_4          conda-forge/linux-64     Cached
  + nltk                     3.6.7  pyhd8ed1ab_0        conda-forge/noarch         1 MB
  + openssl                  3.0.0  h7f98852_2          conda-forge/linux-64     Cached
  + pip                     21.3.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pycodestyle              2.8.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + pyflakes                 2.4.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + python                  3.10.1  h543edf9_2_cpython  conda-forge/linux-64      30 MB
  + python_abi                3.10  2_cp310             conda-forge/linux-64     Cached
  + pyyaml                     6.0  py310h6acc77f_3     conda-forge/linux-64     Cached
  + readline                   8.1  h46c0cb4_0          conda-forge/linux-64     Cached
  + regex               2021.11.10  py310h6acc77f_0     conda-forge/linux-64     Cached
  + setuptools              60.2.0  py310hff52083_0     conda-forge/linux-64     Cached
  + six                     1.16.0  pyh6c4a22f_0        conda-forge/noarch       Cached
  + sqlite                  3.37.0  h9cd32fc_0          conda-forge/linux-64     Cached
  + tk                      8.6.11  h27826a3_1          conda-forge/linux-64     Cached
  + tornado                    6.1  py310h6acc77f_2     conda-forge/linux-64     Cached
  + tqdm                    4.62.3  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + typing_extensions        4.0.1  pyha770c72_0        conda-forge/noarch       Cached
  + tzdata                   2021e  he74cb21_0          conda-forge/noarch       Cached
  + wheel                   0.37.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + xz                       5.2.5  h516909a_1          conda-forge/linux-64     Cached
  + yaml                     0.2.5  h7f98852_2          conda-forge/linux-64      87 KB
  + zipp                     3.6.0  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + zlib                    1.2.11  h36c2ea0_1013       conda-forge/linux-64     Cached

  Summary:

  Install: 48 packages

  Total download: 38 MB
```

</details>

Now that `flake8` comes first, `mamba` grabs `importlib-metadata 3.3.0` (which I believe is related to it being the latest `noarch` version of the package).

I think the "correct" solve here is to fail since `importlib-metadata` isn't a `noarch` package and there's no `py310` package for it, but I'm reporting this only to point out that the constraints for solving appear to be related to the order in which they are being passed in, which seems off.

Happy to provide any other info I can.

**Comment by wolfv** [comment]
I understand that it seems a little confusing that the order of packages "matters" but that's the way the libsolv backtracking solver works. It may pick a given "local" minimum (which is also partially what makes it fast). 

One way to work around it is by giving tighter constraints.

**Comment by wolfv** [comment]
Quite often there is no true way of saying which solution is more correct than the other.
